src/app/modules/verification/verification.controller.ts

```diff
@@ -168,6 +175,81 @@ export const handleGetOtp: RequestHandler<
     })
 }
 
+/**
+ * NOTE: This is exported solely for testing
+ * Generates an otp when a user requests to verify a field.
+ * The current answer is signed, and the signature is also saved in the transaction, with the field id as the key.
+ * @param answer The mobile or email number of the user
+ * @param transactionId The id of the transaction to verify
+ * @param formId The id of the form to verify
+ * @param fieldId The id of the field to verify
+ * @returns 201 when otp generated successfully
+ * @returns 400 when the parameters could not be parsed
+ * @returns 400 when the transaction has expired
+ * @returns 400 when the otp could not be sent via sms
+ * @returns 400 when the otp could not be sent via email
+ * @returns 400 when the provided phone number is not valid
+ * @returns 400 when the field type is not supported for validation
+ * @returns 404 when the requested form was not found
+ * @returns 404 when the transaction was not found
+ * @returns 404 when the field was not found
+ * @returns 422 when the user requested for a new otp without waiting
+ * @returns 500 when the otp could not be hashed
+ * @returns 500 when there is a database error
+ */
+export const _handleGenerateOtp: RequestHandler<
+  { transactionId: string; formId: string; fieldId: string },
+  ErrorDto,
+  { answer: string }
+> = async (req, res) => {
+  const { transactionId, formId, fieldId } = req.params
+  const { answer } = req.body
+  const logMeta = {
+    action: 'handleGenerateOtp',
+    transactionId,
+    fieldId,
+    ...createReqMeta(req),
+  }
+  return FormService.retrieveFormById(formId)
+    .andThen(() => generateOtpWithHash(logMeta, SALT_ROUNDS))
+    .andThen(({ otp, hashedOtp }) =>
+      VerificationFactory.sendNewOtp({
+        fieldId,
+        hashedOtp,
+        otp,
+        recipient: answer,
+        transactionId,
+      }),
+    )
+    .map(() => res.sendStatus(StatusCodes.CREATED))
+    .mapErr((error) => {
+      logger.error({
+        message: 'Error creating new OTP',
+        meta: logMeta,
+        error,
+      })
+      const { errorMessage, statusCode } = mapRouteError(error)
+      return res.status(statusCode).json({ message: errorMessage })
+    })
+}
+
+/**
+ * Handler for the POST /forms/:formId/fieldverifications/:transactionId/fields/:fieldId/otp/generate endpoint
+ */
+export const handleGenerateOtp = [
+  celebrate({
+    [Segments.BODY]: Joi.object({
+      fieldType: Joi.string().valid(...VERIFIED_FIELDTYPES),
```

Comment:
this is not used in the controller right? doesn't seem necessary

Reply:
this is needed for stronger validation - `fieldType` is added as a discriminant because prior to this, we were doing validation but expecting any string to come through. 

since we're migrating to a new endpoint, can add stronger validation so that invalid emails/phone numbers are caught and prevented

Reply:
this isn't present on the controller because it's strictly validation only and hence, is beyond the bounds of what the controller has to know

Reply:
1. that's circular logic - you're adding something to validate and then validating it. this doesn't mean that the validation is important
2. what do you mean by "discriminant" and where were we doing the validation on fieldType before this?
3. if there is genuinely an application-level need for this validation, imo it should happen in the controller. we shouldn't implicitly delegate application logic to the Joi validators. they exist purely so that we can be sure of the types in the controllers.

Reply:
note that I am referring to `fieldType`, not `answer`

Reply:
I see what you mean by "discriminant" now. I'm still a bit hesitant to introduce new validation on public routes, given the many issues we've faced before re: phone number validation. bear in mind that the covid form uses verified fields, and literally the whole population could be using it, so lives are literally at stake. but if you've done your research on the reliability and universality of Joi validation, I'll trust your judgement.

Reply:
i think given what you've said, the best (for now), would be to make the phone number validation as permissive as possible (i.e., defaulting to the previous implementation where any string is allowed), especially because yknow, people are counting on this @___@ but with a TODO / issue raised so that we're cognizant going forward that once this period ends (yes, i still believe covid will end and we'll go back to normalcy), we can correct for stronger phone number validation